Try loading fonts with CTFontManagerCreateMemorySafeFontDescriptorFromData in Lockdown Mode.

apple-fcloutier opened the pull request:
#### dea5f8e4a1495d6a7ae3d295d12ec5a99969861c
<pre>
Try loading fonts with CTFontManagerCreateMemorySafeFontDescriptorFromData in Lockdown Mode.

<a href="https://bugs.webkit.org/show_bug.cgi?id=267948">https://bugs.webkit.org/show_bug.cgi?id=267948</a>

<a href="https://rdar.apple.com/121280542">rdar://121280542</a>

Reviewed by Brent Fulgham and Chris Dumez.

Most Web fonts are not parsed in Lockdown Mode. On systems that provide an
alternative, safe font parser, try to use that font parser. Whether to enable
a fallback font parser is controlled by the LockdownFontParserEnabled setting,
added in &quot;testable&quot; status.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/pal/cf/CoreTextSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreTextSoftLink.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.order:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::appendSources):
* Source/WebCore/css/CSSFontFaceSource.cpp:
(WebCore::CSSFontFaceSource::load):
* Source/WebCore/css/FontFace.cpp:
(WebCore::FontFace::create):
* Source/WebCore/loader/cache/CachedFont.cpp:
(WebCore::CachedFont::policyForCustomFont):
(WebCore::CachedFont::finishLoading):
(WebCore::CachedFont::ensureCustomFontData):
(WebCore::CachedFont::createCustomFontData):
(WebCore::CachedFont::createCustomFontDataExperimentalParser):
(WebCore::CachedFont::shouldAllowCustomFont): Deleted.
* Source/WebCore/loader/cache/CachedFont.h:
(WebCore::CachedFont::didRefuseToParseCustomFont const):
(WebCore::CachedFont::didRefuseToLoadCustomFont const): Deleted.
* Source/WebCore/loader/cache/CachedFontLoadRequest.h:
* Source/WebCore/loader/cache/TrustedFonts.cpp: Renamed from Source/WebCore/loader/cache/AllowedFonts.cpp.
(WebCore::trustedFontHashesInLockdownMode):
(WebCore::hashForFontData):
(WebCore::fontBinaryParsingPolicy):
* Source/WebCore/loader/cache/TrustedFonts.h: Renamed from Source/WebCore/loader/cache/AllowedFonts.h.
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldEnableFontLoadingAPIQuirk const):
* Source/WebCore/page/Settings.yaml:
* Source/WebCore/page/SettingsBase.h:
* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
(WebCore::extractFontCustomPlatformData):
(WebCore::FontCustomPlatformData::create):
(WebCore::FontCustomPlatformData::createMemorySafe):
(WebCore::FontCustomPlatformData::tryMakeFromSerializationData):
(WebCore::createFontCustomPlatformData): Deleted.
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::FontPlatformData::tryMakeFontPlatformData):
* Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
(WebCore::FontCustomPlatformData::create):
(WebCore::FontCustomPlatformData::createMemorySafe):
(WebCore::createFontCustomPlatformData): Deleted.
* Source/WebCore/platform/graphics/win/FontCustomPlatformDataWin.cpp:
(WebCore::FontCustomPlatformData::create):
(WebCore::FontCustomPlatformData::createMemorySafe):
(WebCore::createFontCustomPlatformData): Deleted.
* Source/WebCore/workers/WorkerFontLoadRequest.cpp:
(WebCore::WorkerFontLoadRequest::ensureCustomFontData):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;FontCustomPlatformData&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/win/WebCoreArgumentCodersWin.cpp:
(IPC::ArgumentCoder&lt;Font&gt;::decodePlatformData):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::adjustSettingsForLockdownMode):

Canonical link: <a href="https://commits.webkit.org/274689@main">https://commits.webkit.org/274689@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e8c3d2561361bc073baade9f31cb9a790107e567

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39748 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18727 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42104 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42283 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42055 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21632 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16056 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40322 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15796 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34387 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13692 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35338 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43560 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/33201 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36086 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35673 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39449 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39374 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14563 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11989 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16166 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/46382 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8903 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16230 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9537 "Found 2 new JSC stress test failures: stress/sampling-profiler-microtasks.js.dfg-eager, stress/sampling-profiler-microtasks.js.mini-mode (failure)") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15826 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -2534,6 +2534,19 @@ EventHandlerDrivenSmoothKeyboardScrollingEnabled:
     WebCore:
       default: false
 
+ExperimentalFontParserInLockdownMode:
```
I think we shouldn't name it "Experimental".
All the flags here are already experimental.

brentfulgham replied:
Right -- let's call it `LockdownFontParserEnabled`

vitorroriz commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -2534,6 +2534,19 @@ EventHandlerDrivenSmoothKeyboardScrollingEnabled:
 
+ExperimentalFontParserInLockdownMode:
+  type: bool
+  status: embedder
```
`testable` would be a better status here, because we intend to turn it on by default at some point when this is mature enough. Also, this will allow the feature to be testable in our pipeline bots.

brentfulgham replied:
Agreed.

vitorroriz commented on Source/WebCore/loader/cache/TrustedFonts.h:
```diff
@@ -31,17 +31,18 @@
+// using the default, possibly-unsafe font parser.
 // Any: any font binary will be downloaded, no checks will be done during load.
-// Restricted: any font binary will be donwloaded but just binaries listed by WebKit will be allowed loaded. A not allowed binary will be deleted after the check is done (see CachedFont::shouldAllowCustomFont()).
+// Restricted: any font binary will be donwloaded but just binaries listed by WebKit are trusted to load through the system font parser. A not allowed binary will be deleted after the check is done (see CachedFont::shouldTrustCustomFont()). If the system advertises a memory-safe font parser, it will be used as a fallback for fonts that are not trusted.
```
```suggestion
// Restricted: any font binary will be donwloaded but just binaries listed by WebKit are trusted to load through the system font parser. A not trusted binary will be deleted after the check is done (see CachedFont::shouldTrustCustomFont()). If the system advertises a memory-safe font parser, it will be used as a fallback for fonts that are not trusted.
```

vitorroriz commented on Source/WebCore/loader/cache/CachedFont.cpp:
```diff
@@ -134,8 +129,16 @@ String CachedFont::calculateItemInCollection() const
-        m_fontCustomPlatformData = createCustomFontData(*data, calculateItemInCollection(), wrapping);
+        bool wrapping = false;
+        if (m_mustLoadWithMemorySafeParser) {
+            if (m_loader->frame()->settings().experimentalFontParserInLockdownMode())
```
I think we should avoid the 'Experimental' name here as I mentioned in `UnifiedWebPreferences`

vitorroriz commented on Source/WebCore/loader/cache/CachedFont.cpp:
```diff
@@ -134,8 +129,16 @@ String CachedFont::calculateItemInCollection() const
+                m_fontCustomPlatformData = createCustomFontDataExperimentalParser(*data, calculateItemInCollection(), wrapping);
+            if (!m_fontCustomPlatformData) {
+                m_didRefuseToParseCustomFont = true;
+                setErrorAndDeleteData();
```
It is just that it is a bit weird for me that we are trying to parse at `CachedFont::ensureCustomFontData(SharedBuffer* data)` and an ensure function can potentially still invalidate loading. At that point I’d expect loading to be concluded already, and parsing would be part of loading.

Notice that `ensureCustomFontData` is called from `CSSFontFaceSource::fontLoaded`

Maybe we can come up with a better abstraction for it? Otherwise is not critical, of course.

apple-fcloutier replied:
FWIW, parsing was done in `ensureCustomFontData` before my change (see [138-](https://github.com/WebKit/WebKit/pull/23123/files#diff-47a745755cc9d17fbba3686ad6e0eadc2a4bb514158235a72e81389552fcc04eL138)). I agree that `setErrorAndDeleteData` will probably never actually cancel the m_loader.

vitorroriz commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4438,14 +4438,15 @@ static void adjustSettingsForLockdownMode(Settings& settings, const WebPreferenc
+    case DownloadableBinaryFontTrustedTypes::None:
         break;
     }
+    settings.setExperimentalFontParserInLockdownMode(true);
```
Why are we setting this to true here unconditionally? Are we trying to override the default behavior?

apple-fcloutier replied:
I think I don't understand the enablement strategy for Lockdown Mode features. Right now there are 2 states:

* Lockdown Mode is off, fonts are parsed through the system font parser
* Lockdown Mode is on, fonts on the allowlist are parsed through the system font parser and the rest are parsed through the experimental font parser

Should I have another feature flag/another feature level so that we also get a third state? "Lockdown Mode is on, but the experimental font parser is disabled"

cdumez commented on Source/WebCore/loader/cache/CachedFont.cpp:
```diff
@@ -147,8 +150,21 @@ bool CachedFont::ensureCustomFontData(SharedBuffer* data)
+RefPtr<FontCustomPlatformData> CachedFont::createCustomFontDataExperimentalParser(SharedBuffer& bytes, const String& itemInCollection, bool& wrapping)
+{
+    RefPtr buffer = { &bytes };
+    if (!buffer)
```
This null check makes no sense since you got a C++ reference as parameter. Also, you don't need to store bytes in a smart pointer, it is up to the caller of createCustomFontDataExperimentalParser() to ref this object.

cdumez commented on Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
```diff
@@ -100,6 +108,23 @@ RefPtr<FontCustomPlatformData> createFontCustomPlatformData(SharedBuffer& buffer
+        return nullptr;
+    }
+    auto fontDescriptor = adoptCF(PAL::softLinkCoreTextCTFontManagerCreateMemorySafeFontDescriptorFromData(extractedData.get()));
+    auto protectedBuffer = SharedBuffer::create(extractedData.get());
```
just `buffer` is enough, no need to mention it is "protected". Also, please replace `auto` with `Ref` to make it clear we're dealing with a smart pointer here (same advice likely applies elsewhere in this PR).

cdumez commented on Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
```diff
@@ -100,6 +108,23 @@ RefPtr<FontCustomPlatformData> createFontCustomPlatformData(SharedBuffer& buffer
     return adoptRef(new FontCustomPlatformData(fontDescriptor.get(), WTFMove(creationData)));
 }
 
+RefPtr<FontCustomPlatformData> createFontCustomPlatformDataMemorySafe(SharedBuffer& buffer, const String& itemInCollection)
```
Why isn't this a static function on `FontCustomPlatformData` ? `FontCustomPlatformData::createMemorySafe()`

apple-fcloutier replied:
The boring reason is that I put it in the same scope as `createFontCustomPlatformData`. I can make both static functions.

apple-fcloutier replied:
Hm, I don't know if I should do that. `createFontCustomPlatformData` is `WEBCORE_EXPORT`. Am I going to break something/someone if I change its ABI name? If I leave it there, should I still move `createFontCustomPlatformDataMemorySafe`?

cdumez replied:
WebCore is only used by WebKit there is no ABI compatibility issue. We own all the callers, in the same codebase.

brentfulgham reviewed (commented):
This patch was made significantly more complicated by deciding to do a mass rename of "Allowed" fonts to "Trusted" fonts. It would be much easier to move this forward if we restricted one patch to the mechanical rename, and a second patch for the specific logic changes.

brentfulgham commented on Source/WebCore/loader/cache/CachedFont.cpp:
```diff
@@ -147,8 +150,21 @@ bool CachedFont::ensureCustomFontData(SharedBuffer* data)
 RefPtr<FontCustomPlatformData> CachedFont::createCustomFontData(SharedBuffer& bytes, const String& itemInCollection, bool& wrapping)
 {
     RefPtr buffer = { &bytes };
+    if (!buffer)
```
I don't think it's possible for `buffer` to be nullptr, since it is a RefPtr constructed from a reference.

brentfulgham replied:
In fact, if you are just trying to control the lifetime of this SharedBuffer, this should just be a Ref:

`Ref buffer = { bytes };`

cdumez replied:
Same issue I commented on in https://github.com/WebKit/WebKit/pull/23123#discussion_r1476710787

Note that protecting a parameter like this (`Ref buffer = { bytes };`) goes against our smart pointer guidelines. It is up to the caller to ref the buffers before passing it as parameter. The function implementation should NOT have to ref it itself.

apple-fcloutier replied:
Point taken that it can't be nullptr. For `RefPtr` usage, I'm more anxious because the code was already there and I'm not sure that I understand the implications of changing it. Do I need to check callers and potentially change stuff up?

cdumez replied:
Yes, you'd have to check all caller to make sure they old a ref on the stack to the buffer they're passing as parameter to the function. If you don't want to do that, it's OK, you can just use `Ref buffer = { bytes };` here like Brent suggested.

apple-fcloutier replied:
Vitor helped me catch up with smart pointers/references. The reason this is a RefPtr is that convertWOFFToSfntIfNecessary wants a RefPtr. It's probably not all that beneficial to move to Ref as a result: we'd just be converting immediately.

brentfulgham commented on Source/WebCore/loader/cache/CachedFontLoadRequest.h:
```diff
@@ -60,7 +60,16 @@ class CachedFontLoadRequest final : public FontLoadRequest, public CachedFontCli
+        if (!result && m_font->didRefuseToParseCustomFont() && m_context) {
+            auto message = makeString("[Lockdown Mode] This font has been blocked: ", m_font->url().string());
+            m_context->addConsoleMessage(MessageSource::Security, MessageLevel::Info, message);
+        }
```
Nit: This is not new with your patch, but all of this is a lot of complexity to put in a header file. We should make a `CachedFontLoadRequest.cpp` file to hold this kind of stuff.

brentfulgham reviewed (changes requested):
I think this is a great first cut, but I think we need a bit of cleanup before we can merge this.

cdumez commented on Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
```diff
@@ -88,15 +89,37 @@ RefPtr<FontCustomPlatformData> createFontCustomPlatformData(SharedBuffer& buffer
+
+RefPtr<FontCustomPlatformData> FontCustomPlatformData::create(SharedBuffer& buffer, const String& itemInCollection)
+{
+    auto extractedData = extractFontCustomPlatformData(buffer, itemInCollection);
```
`auto` -> `RetainPtr`. We prefer not to hide smart pointers behind `auto` in new code.

cdumez commented on Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
```diff
@@ -88,15 +89,37 @@ RefPtr<FontCustomPlatformData> createFontCustomPlatformData(SharedBuffer& buffer
+        // Something is wrong with the font.
+        return nullptr;
+    }
+    RetainPtr<CTFontDescriptorRef> fontDescriptor = adoptCF(CTFontManagerCreateFontDescriptorFromData(extractedData.get()));
```
is `<CTFontDescriptorRef>` really needed here?

cdumez commented on Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
```diff
@@ -88,15 +89,37 @@ RefPtr<FontCustomPlatformData> createFontCustomPlatformData(SharedBuffer& buffer
+        return nullptr;
+    }
+    RetainPtr<CTFontDescriptorRef> fontDescriptor = adoptCF(CTFontManagerCreateFontDescriptorFromData(extractedData.get()));
+    Ref<SharedBuffer> bufferRef = SharedBuffer::create(extractedData.get());
```
`<SharedBuffer>` is not needed.

cdumez commented on Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
```diff
@@ -88,15 +89,37 @@ RefPtr<FontCustomPlatformData> createFontCustomPlatformData(SharedBuffer& buffer
+    if (!PAL::canLoad_CoreText_CTFontManagerCreateMemorySafeFontDescriptorFromData())
+        return nullptr;
+
+    auto extractedData = extractFontCustomPlatformData(buffer, itemInCollection);
```
`auto` -> `RetainPtr`

cdumez commented on Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
```diff
@@ -88,15 +89,37 @@ RefPtr<FontCustomPlatformData> createFontCustomPlatformData(SharedBuffer& buffer
     }
-    auto fontDescriptor = adoptCF(CTFontManagerCreateFontDescriptorFromData(extractedData.get()));
-    auto protectedBuffer = SharedBuffer::create(extractedData.get());
+    RetainPtr<CTFontDescriptorRef> fontDescriptor = adoptCF(PAL::softLinkCoreTextCTFontManagerCreateMemorySafeFontDescriptorFromData(extractedData.get()));
```
Is `<CTFontDescriptorRef>` really needed here?

cdumez commented on Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
```diff
@@ -88,15 +89,37 @@ RefPtr<FontCustomPlatformData> createFontCustomPlatformData(SharedBuffer& buffer
-    auto fontDescriptor = adoptCF(CTFontManagerCreateFontDescriptorFromData(extractedData.get()));
-    auto protectedBuffer = SharedBuffer::create(extractedData.get());
+    RetainPtr<CTFontDescriptorRef> fontDescriptor = adoptCF(PAL::softLinkCoreTextCTFontManagerCreateMemorySafeFontDescriptorFromData(extractedData.get()));
+    Ref<SharedBuffer> bufferRef = SharedBuffer::create(extractedData.get());
```
Please omit `<SharedBuffer>`

cdumez reviewed (commented):
I'm not a font person but this looks much better. I made a few nits comments. The commit log needs fixing too.

cdumez reviewed (commented):
The commit log should say:
```
Reviewed by NOBODY (OOPS!).
```

The bots are in charge of updating that line with the proper reviewers once the PR gets merged.

brentfulgham commented on Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
```diff
@@ -88,22 +89,44 @@ RefPtr<FontCustomPlatformData> createFontCustomPlatformData(SharedBuffer& buffer
+    if (!extractedData) {
+        // Something is wrong with the font.
+        return nullptr;
+    }
```
Nit: Should have a newline here.

brentfulgham commented on Source/WebCore/platform/graphics/win/FontCustomPlatformDataWin.cpp:
```diff
@@ -100,6 +100,11 @@ RefPtr<FontCustomPlatformData> createFontCustomPlatformData(SharedBuffer& buffer
     return adoptRef(new FontCustomPlatformData(fontName, WTFMove(creationData)));
```
It seems like we could be doing a better job of avoiding the copying of font buffer data in some of these create methods. (Not necessary for this patch -- just an idea for improvement).

brentfulgham reviewed (commented):
I think this looks good, but I'm going to ask Chris and Vitor to confirm their comments were fully addressed before I do a final sign-off.

cdumez reviewed (approved):
LGTM

brentfulgham reviewed (approved):
I confirmed with Chris and Vitor that they approve, as well.

apple-fcloutier committed:
Try loading fonts with CTFontManagerCreateMemorySafeFontDescriptorFromData in Lockdown Mode.

https://bugs.webkit.org/show_bug.cgi?id=267948

rdar://121280542

Reviewed by Brent Fulgham and Chris Dumez.

Most Web fonts are not parsed in Lockdown Mode. On systems that provide an
alternative, safe font parser, try to use that font parser. Whether to enable
a fallback font parser is controlled by the LockdownFontParserEnabled setting,
added in "testable" status.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/pal/cf/CoreTextSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreTextSoftLink.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.order:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::appendSources):
* Source/WebCore/css/CSSFontFaceSource.cpp:
(WebCore::CSSFontFaceSource::load):
* Source/WebCore/css/FontFace.cpp:
(WebCore::FontFace::create):
* Source/WebCore/loader/cache/CachedFont.cpp:
(WebCore::CachedFont::policyForCustomFont):
(WebCore::CachedFont::finishLoading):
(WebCore::CachedFont::ensureCustomFontData):
(WebCore::CachedFont::createCustomFontData):
(WebCore::CachedFont::createCustomFontDataExperimentalParser):
(WebCore::CachedFont::shouldAllowCustomFont): Deleted.
* Source/WebCore/loader/cache/CachedFont.h:
(WebCore::CachedFont::didRefuseToParseCustomFont const):
(WebCore::CachedFont::didRefuseToLoadCustomFont const): Deleted.
* Source/WebCore/loader/cache/CachedFontLoadRequest.h:
* Source/WebCore/loader/cache/TrustedFonts.cpp: Renamed from Source/WebCore/loader/cache/AllowedFonts.cpp.
(WebCore::trustedFontHashesInLockdownMode):
(WebCore::hashForFontData):
(WebCore::fontBinaryParsingPolicy):
* Source/WebCore/loader/cache/TrustedFonts.h: Renamed from Source/WebCore/loader/cache/AllowedFonts.h.
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldEnableFontLoadingAPIQuirk const):
* Source/WebCore/page/Settings.yaml:
* Source/WebCore/page/SettingsBase.h:
* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
(WebCore::extractFontCustomPlatformData):
(WebCore::FontCustomPlatformData::create):
(WebCore::FontCustomPlatformData::createMemorySafe):
(WebCore::FontCustomPlatformData::tryMakeFromSerializationData):
(WebCore::createFontCustomPlatformData): Deleted.
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::FontPlatformData::tryMakeFontPlatformData):
* Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
(WebCore::FontCustomPlatformData::create):
(WebCore::FontCustomPlatformData::createMemorySafe):
(WebCore::createFontCustomPlatformData): Deleted.
* Source/WebCore/platform/graphics/win/FontCustomPlatformDataWin.cpp:
(WebCore::FontCustomPlatformData::create):
(WebCore::FontCustomPlatformData::createMemorySafe):
(WebCore::createFontCustomPlatformData): Deleted.
* Source/WebCore/workers/WorkerFontLoadRequest.cpp:
(WebCore::WorkerFontLoadRequest::ensureCustomFontData):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<FontCustomPlatformData>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/win/WebCoreArgumentCodersWin.cpp:
(IPC::ArgumentCoder<Font>::decodePlatformData):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::adjustSettingsForLockdownMode):

Canonical link: https://commits.webkit.org/274689@main
